### PR TITLE
Code splitting ready for prime time

### PR DESCRIPTION
**Pull request by markATavail**
Splitting is stable, and has run the Avail tests thousands of times.  There is
a glitch that happens ever few hundred runs related to `Acquire_` on an
Avail mutex, but it's unclear when this problem surfaced.

### CODE REVIEW

File: avail/src/main/kotlin/avail/interpreter/levelTwo/L2Instruction.kt [hunk 1]
```diff
@@ -477,4 +478,7 @@ constructor(
 			operand.postOptimizationCleanup()
 		}
 	}
+
+	fun interestingConditions(): List<L2SplitCondition> =
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/interpreter/levelTwo/L2Instruction.kt]: This really needs to be documented. It's practically criminal that I don't know what's interesting about them. Seems really important to the narrative.

File: avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_GREATER_THAN_CONSTANT.kt [hunk 2]
```diff
@@ -103,4 +106,20 @@ object L2_JUMP_IF_GREATER_THAN_CONSTANT : L2ConditionalJump(
 			ifGreater,
 			ifNotGreater)
 	}
+
+	//TODO override emit...
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_GREATER_THAN_CONSTANT.kt]: Is this `TODO` concerning?

File: avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_GREATER_THAN_OBJECT.kt [hunk 3]
```diff
@@ -102,4 +105,23 @@ object L2_JUMP_IF_GREATER_THAN_OBJECT : L2ConditionalJump(
 			ifGreater,
 			ifNotGreater)
 	}
+
+	//TODO override emit...
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_GREATER_THAN_OBJECT.kt]: Is this `TODO` concerning?

File: avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_GREATER_THAN_OR_EQUAL_TO_CONSTANT.kt [hunk 4]
```diff
@@ -104,4 +107,20 @@ object L2_JUMP_IF_GREATER_THAN_OR_EQUAL_TO_CONSTANT : L2ConditionalJump(
 			ifGreaterOrEqual,
 			ifNotGreaterOrEqual)
 	}
+
+	//TODO override emit...
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_GREATER_THAN_OR_EQUAL_TO_CONSTANT.kt]: Is this `TODO` concerning?

File: avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_GREATER_THAN_OR_EQUAL_TO_OBJECT.kt [hunk 5]
```diff
@@ -102,4 +105,23 @@ object L2_JUMP_IF_GREATER_THAN_OR_EQUAL_TO_OBJECT : L2ConditionalJump(
 			ifGreaterOrEqual,
 			ifNotGreaterOrEqual)
 	}
+
+	//TODO override emit...
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_GREATER_THAN_OR_EQUAL_TO_OBJECT.kt]: Is this `TODO` concerning?

File: avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_LESS_THAN_CONSTANT.kt [hunk 6]
```diff
@@ -98,4 +101,20 @@ object L2_JUMP_IF_LESS_THAN_CONSTANT : L2ConditionalJump(
 		emitBranch(
 			translator, method, instruction, Opcodes.IFNE, ifLess, ifNotLess)
 	}
+
+	//TODO override emit...
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_LESS_THAN_CONSTANT.kt]: Is this `TODO` concerning?

File: avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_LESS_THAN_OBJECT.kt [hunk 7]
```diff
@@ -102,4 +105,23 @@ object L2_JUMP_IF_LESS_THAN_OBJECT : L2ConditionalJump(
 			ifLess,
 			ifNotLess)
 	}
+
+	//TODO override emit...
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_LESS_THAN_OBJECT.kt]: Is this `TODO` concerning?

File: avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_LESS_THAN_OR_EQUAL_TO_CONSTANT.kt [hunk 8]
```diff
@@ -104,4 +107,20 @@ object L2_JUMP_IF_LESS_THAN_OR_EQUAL_TO_CONSTANT : L2ConditionalJump(
 			ifLessOrEqual,
 			ifNotLessOrEqual)
 	}
+
+	//TODO override emit...
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_LESS_THAN_OR_EQUAL_TO_CONSTANT.kt]: Is this `TODO` concerning?

File: avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_LESS_THAN_OR_EQUAL_TO_OBJECT.kt [hunk 9]
```diff
@@ -102,4 +105,23 @@ object L2_JUMP_IF_LESS_THAN_OR_EQUAL_TO_OBJECT : L2ConditionalJump(
 			ifLessOrEqual,
 			ifNotLessOrEqual)
 	}
+
+	//TODO override emit...
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_LESS_THAN_OR_EQUAL_TO_OBJECT.kt]: Is this `TODO` concerning?

File: avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_OBJECTS_EQUAL.kt [hunk 10]
```diff
@@ -112,4 +115,23 @@ object L2_JUMP_IF_OBJECTS_EQUAL : L2ConditionalJump(
 		emitBranch(
 			translator, method, instruction, Opcodes.IFNE, ifEqual, ifNotEqual)
 	}
+
+	//TODO override emit...
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_OBJECTS_EQUAL.kt]: Is this `TODO` concerning?

File: avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_UNBOX_INT.kt [hunk 11]
```diff
@@ -126,4 +131,70 @@ object L2_JUMP_IF_UNBOX_INT : L2ConditionalJump(
...
+		val generator = regenerator.targetGenerator
+		val manifest = generator.currentManifest
+		val restriction = manifest.restrictionFor(source.semanticValue())
+		// See if there's an int version of a synonym of the source.  There
+		// must be a less messy way of doing this.
+		val sourceInt = manifest.semanticValueToSynonym(source.semanticValue())
+			.semanticValues()
+			.map(::L2SemanticUnboxedInt)
+			.firstOrNull(manifest::hasSemanticValue)
+			?: L2SemanticUnboxedInt(source.semanticValue())
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_UNBOX_INT.kt:153]: Based on the comment, maybe this needs to be factored into a named operation? Or is this the only place where you need to do this?

File: avail/src/main/kotlin/avail/optimizer/L2Optimizer.kt [hunk 12]
```diff
@@ -222,6 +224,194 @@ class L2Optimizer internal constructor(
...
+									condition.holdsFor(edge.manifest()) -> '+'
+									else -> '-'
+								})
+						}
+						block.debugNote.append(' ')
+						block.debugNote.append(condition)
+					}
+				}
+			}
+		}
```
Comment (toddATavail) [inline avail/src/main/kotlin/avail/optimizer/L2Optimizer.kt]: This looks like it should be removed or conditionalized in some way.

**Review by toddATavail — CHANGES_REQUESTED** [review]
I had small quibbles with your comments (or lack thereof) occasionally, and questions about some TODOs, but your techniques and logic look solid. Glad we have tests to prove general correctness, though! Would be extraordinarily hard by eye!

File: avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_OBJECTS_EQUAL.kt (hunk 10 above)
Comment (markATavail) [reply avail/src/main/kotlin/avail/interpreter/levelTwo/operation/L2_JUMP_IF_OBJECTS_EQUAL.kt]: This and some related comments have been obviated by completely rewriting the conditional branch instruction hierarchy, specifically introducing NumericCompare as a helper.

**Review by toddATavail — APPROVED** [review]
There's a lot here, but I glanced at most everything. I will have to take your word (and test results) that everything is ready. We ought to fix the 2x performance regression, though; it certainly wasn't the goal for code splitting to be a _de_-optimization.

**Pull request merged**